src/Microsoft.OData.Core/ODataUtils.cs

```diff
@@ -206,11 +206,11 @@ public static ODataServiceDocument GenerateServiceDocument(this IEdmModel model)
 
             ODataServiceDocument serviceDocument = new ODataServiceDocument();
             serviceDocument.EntitySets = model.EntityContainer.EntitySets()
-                .Select(entitySet => new ODataEntitySetInfo() { Name = entitySet.Name, Title = entitySet.Name, Url = new Uri(entitySet.Name, UriKind.RelativeOrAbsolute) }).ToList();
+                .Select(entitySet => new ODataEntitySetInfo() { Name = entitySet.Name, Title = entitySet.Name, Url = new Uri(entitySet.Name, UriKind.RelativeOrAbsolute) });
             serviceDocument.Singletons = model.EntityContainer.Singletons()
-                .Select(singleton => new ODataSingletonInfo() { Name = singleton.Name, Title = singleton.Name, Url = new Uri(singleton.Name, UriKind.RelativeOrAbsolute) }).ToList();
+                .Select(singleton => new ODataSingletonInfo() { Name = singleton.Name, Title = singleton.Name, Url = new Uri(singleton.Name, UriKind.RelativeOrAbsolute) });
             serviceDocument.FunctionImports = model.EntityContainer.OperationImports().OfType<IEdmFunctionImport>().Where(functionImport => functionImport.IncludeInServiceDocument && !functionImport.Function.Parameters.Any())
-                .Select(functionImport => new ODataFunctionImportInfo() { Name = functionImport.Name, Title = functionImport.Name, Url = new Uri(functionImport.Name, UriKind.RelativeOrAbsolute) }).ToList();
+                .Select(functionImport => new ODataFunctionImportInfo() { Name = functionImport.Name, Title = functionImport.Name, Url = new Uri(functionImport.Name, UriKind.RelativeOrAbsolute) });

```

Comment:
If you remove the `ToList()`, does that mean each time the caller calls `serviceDocument.EntitySets` or `serviceDocument.Singletons` it will re-evaluate the query? Wouldn't that be a penalty if those properties are accessed frequently? Or is the rationale to make it the responsibility of the caller to call `ToList()` on the property if they intend to iterate over them multiple times?

Reply:
That's the trade-off.  If we only iterate once (or not at all) than the enumerable is more performant.  But, if we may iterate multiple times then it's worth building the list.  Lazyloading the lists falls somewhere in between -- you still have the benefit of the list for multiple iterations, but not the overhead if nobody ever asks for it.

@Sreejith -- does AGS ever call this method? If so, is it called just once or multiple times? How does WebAPI use it?

---
In reply to: [570816077](https://github.com/OData/odata.net/pull/1993#discussion_r570816077) [](ancestors = 570816077)

Reply:
Dont see it used by WebAPI, have asked JP whether thy use it in AGS

Reply:
Asked JP , he told its used only once, so better to remove TOList()